Implements `unmanaged-cluster` profiles

jpmcb opened the pull request:
- Users can specify --profile via flag or config to automatically
  install a profile from an installed package repo.
- Users can set the --profile-version to install a specific version.
  This is optional and will default the the latest semver if not
  specified
- Users can set the --profile-config-file for a values yaml to be
  consumed by the installed profile
- Also includes a change to logging of packages being reconciled
  so the name is visible to users when reconciling them
- Updates our usage function for cobra so a nicely formatted line break
  for long flag messages don't wrap around

## Which issue(s) this PR fixes
Fixes: #3537

## Describe testing done for PR

Install a profile by name:

```
$ tanzu unmanaged-cluster create test --profile fluent-bit.community.tanzu.vmware.com

📁 Created cluster directory

🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Compatibility file exists at /home/jmcb/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v3

🔧 Resolving TKr
   projects.registry.vmware.com/tce/tkr:v0.17.0-dev-1
   TKr exists at /home/jmcb/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v0.17.0-dev-1
   Rendered Config: /home/jmcb/.config/tanzu/tkg/unmanaged/test/config.yaml
   Bootstrap Logs: /home/jmcb/.config/tanzu/tkg/unmanaged/test/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind:v1.22.4

📦 Selected core package repository
   projects.registry.vmware.com/tce/repo-10:0.10.0

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.11.0

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tce/kapp-controller-multi-pkg:v0.30.1

🚀 Creating cluster test
   Cluster creation using kind!
   ❤️  Checkout this awesome project at https://kind.sigs.k8s.io
   Base image downloaded
   Cluster created
   To troubleshoot, use:

   kubectl ${COMMAND} --kubeconfig /home/jmcb/.config/tanzu/tkg/unmanaged/test/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   tkg-core-repository package repo status: Reconcile succeeded
   projects.registry.vmware.com-tce-main-0.11.0 package repo status: Reconcile succeeded

🌐 Installing CNI
   calico.community.tanzu.vmware.com:3.22.1

🌐 Installing Profile fluent-bit.community.tanzu.vmware.com
   Installing profile without version specified. Using version 1.7.5
   Installing profile with no configuration file

✅ Cluster created

🎮 kubectl context set to test

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete test

$ k get pods -A
NAMESPACE            NAME                                         READY   STATUS    RESTARTS   AGE
fluent-bit           fluent-bit-k8xq6                             0/1     Running   0          51s
kube-system          calico-kube-controllers-556b886dc8-4zc4q     1/1     Running   0          48s
kube-system          calico-node-94s4j                            1/1     Running   0          48s
kube-system          coredns-78fcd69978-cxzbd                     1/1     Running   0          95s
kube-system          coredns-78fcd69978-hkknt                     1/1     Running   0          95s
kube-system          etcd-test-control-plane                      1/1     Running   0          108s
kube-system          kube-apiserver-test-control-plane            1/1     Running   0          108s
kube-system          kube-controller-manager-test-control-plane   1/1     Running   0          108s
kube-system          kube-proxy-hj9ns                             1/1     Running   0          95s
kube-system          kube-scheduler-test-control-plane            1/1     Running   0          108s
local-path-storage   local-path-provisioner-85494db59d-bf9ws      1/1     Running   0          95s
tkg-system           kapp-controller-779d9777dc-k8xm7             1/1     Running   0          95s

$ k get packageinstalls -n tkg-system
NAME                                    PACKAGE NAME                            PACKAGE VERSION   DESCRIPTION           AGE
cni                                     calico.community.tanzu.vmware.com       3.22.1            Reconcile succeeded   64s
fluent-bit.community.tanzu.vmware.com   fluent-bit.community.tanzu.vmware.com   1.7.5             Reconcile succeeded   64s
```

Users may also provide _just_ the prefix of the package name and the install will do a best effort to match the prefix of a package:
```
$ tanzu unmanaged-cluster create --profile fluent-bit
...

🌐 Installing Profile fluent-bit
   Installing profile without version specified. Using version 1.7.5
   Installing profile with no configuration file

...

❯ k get packageinstalls -A
NAMESPACE    NAME         PACKAGE NAME                            PACKAGE VERSION   DESCRIPTION   AGE
tkg-system   cni          calico.community.tanzu.vmware.com       3.22.1            Reconciling   15s
tkg-system   fluent-bit   fluent-bit.community.tanzu.vmware.com   1.7.5             Reconciling   15s
```

Install a profile and specify the version in the profile mapping:
```
$ tanzu unmanaged-cluster create test --profile external-dns.community.tanzu.vmware.com:0.8.0
...

🌐 Installing Profile external-dns.community.tanzu.vmware.com
   Using profile version 0.8.0
   Installing profile with no configuration file

...
```

Selects the latest version if no version tag specified. (note, this includes some code from golang/x/mod/semver to parse semantic versions)
```
$ tanzu unmanaged-cluster create test --profile external-dns.community.tanzu.vmware.com
...

🌐 Installing Profile external-dns.community.tanzu.vmware.com
   Installing profile without version specified. Using version 0.10.0
   Installing profile with no configuration file

...
```

The `latest` keyword is also reserved and will select the latest semver tag
```
$ tanzu unmanaged-cluster create test --profile external-dns.community.tanzu.vmware.com:latest
...

🌐 Installing Profile external-dns.community.tanzu.vmware.com
   Installing profile without version specified. Using version 0.10.0
   Installing profile with no configuration file

...
```


If also using a config, users can use `::` to skip defining a version and select the latest one (or the `latest` keyword)
```
$ tanzu unmanaged-cluster create test --profile external-dns.community.tanzu.vmware.com::my-config.yaml
...

🌐 Installing Profile external-dns.community.tanzu.vmware.com
   Installing profile without version specified. Using version 0.10.0
   Using config my-config.yaml

...
```

Include a values yaml file to be consumed by the package profile. [Using external DNS as an example here](https://github.com/vmware-tanzu/community-edition/tree/main/addons/packages/external-dns/0.10.0)
```yaml
---
#! The namespace in which to deploy ExternalDNS.
namespace: my-cool-namespace

#! Deployment related configuration
deployment:
  args:
  - --source=service
  - --source=contour-httpproxy
  - --txt-owner-id=k8s
  - --domain-filter=k8s.example.org
  - --namespace=my-services-ns
  - --provider=rfc2136
  - --rfc2136-host=100.69.97.77
  - --rfc2136-port=53
  - --rfc2136-zone=k8s.example.org
  - --rfc2136-tsig-secret=MTlQs3NNU=
  - --rfc2136-tsig-secret-alg=hmac-sha256
  - --rfc2136-tsig-keyname=externaldns-key
  - --rfc2136-tsig-axfr
  env: []
  securityContext: []
  volumeMounts: []
  volumes: []

#! Service account related configuration
serviceaccount:
  annotations:
    key: value
```

Using the values yaml file:
```
$ tanzu unmanaged-cluster create test \
  --profile external-dns.community.tanzu.vmware.com:0.10.0:values.yaml

...

🌐 Installing Profile external-dns.community.tanzu.vmware.com
   Using profile version 0.10.0
   Using config values.yaml

...

$ k get pods -A
NAMESPACE            NAME                                         READY   STATUS    RESTARTS      AGE
kube-system          calico-kube-controllers-6849b5c7d-c5w9z      1/1     Running   0             6m5s
kube-system          calico-node-cht4q                            1/1     Running   0             6m5s
kube-system          coredns-78fcd69978-4csnt                     1/1     Running   0             6m55s
kube-system          coredns-78fcd69978-tjsng                     1/1     Running   0             6m55s
kube-system          etcd-test-control-plane                      1/1     Running   0             7m8s
kube-system          kube-apiserver-test-control-plane            1/1     Running   0             7m8s
kube-system          kube-controller-manager-test-control-plane   1/1     Running   0             7m8s
kube-system          kube-proxy-wrrfk                             1/1     Running   0             6m55s
kube-system          kube-scheduler-test-control-plane            1/1     Running   0             7m8s
local-path-storage   local-path-provisioner-85494db59d-c9nj5      1/1     Running   0             6m55s
my-cool-namespace    external-dns-557c9c6487-7d9q5                1/1     Running   4 (50s ago)   6m8s
tkg-system           kapp-controller-779d9777dc-sq6pb             1/1     Running   0             6m55s

$ k get packageinstalls -n tkg-system
NAME                                      PACKAGE NAME                              PACKAGE VERSION   DESCRIPTION           AGE
cni                                       calico.community.tanzu.vmware.com         3.22.1            Reconcile succeeded   6m33s
external-dns.community.tanzu.vmware.com   external-dns.community.tanzu.vmware.com   0.10.0            Reconcile succeeded   6m33s
```

Users can specify multiple profiles via multiple flags
```
$ tanzu unmanaged-cluster create test --profile external-dns.community.tanzu.vmware.com --profile fluent-bit.community.tanzu.vmware.com

...

🌐 Installing Profile external-dns.community.tanzu.vmware.com
   Using profile version 0.10.0
   Using config values.yaml

🌐 Installing Profile fluent-bit.community.tanzu.vmware.com
   Installing profile without version specified. Using version 1.7.5
   Installing profile with no configuration file
```

Users can use a multiple profile mappings delimitted by a `,`
```shell
# this will generate 3 profiles:
# - external-dns version 0.10.0 using the external-values.yaml file
# - a fluent-bit profile at the latest version with no config file
# - the app toolkit, version 0.1.0 with the values.yaml file
# Each profile is delimited by a `,` and each value is delimited by `:`
$ tanzu unmanaged-cluster create test \
  --profile external-dns.community.tanzu.vmware.com:0.10.0:external-values.yaml,fluent-bit.community.tanzu.vmware.com,app-toolkit.community.tanzu.vmware.com:0.1.0:values.yaml
```
results in:
```
🌐 Installing Profile external-dns.community.tanzu.vmware.com
   Using profile version 0.10.0
   Using config external-values.yaml

🌐 Installing Profile fluent-bit.community.tanzu.vmware.com
   Installing profile without version specified. Using version 1.7.5
   Installing profile with no configuration file

🌐 Installing Profile app-toolkit.community.tanzu.vmware.com
   Using profile version 0.1.0
   Using config values.yaml
```

Users can specify profiles in their config file:
```yaml
ClusterName: test
KubeconfigPath: ""
ExistingClusterKubeconfig: ""
NodeImage: ""
Provider: kind
ProviderConfiguration: {}
Cni: calico
CniConfiguration: {}
PodCidr: 10.244.0.0/16
ServiceCidr: 10.96.0.0/16
TkrLocation: ""
AdditionalPackageRepos: []
PortsToForward: []
SkipPreflight: false
ControlPlaneNodeCount: "1"
WorkerNodeCount: "0"
Profiles:
- name: fluent-bit.community.tanzu.vmware.com
- name: external-dns.community.tanzu.vmware.com
  config: external-values.yaml
  version: 0.10.0
- name: app-toolkit.community.tanzu.vmware.com
  config: values.yaml
```
and during deploy:
```
$ tanzu unmanaged-cluster create -f test.yaml
...

🌐 Installing Profile fluent-bit.community.tanzu.vmware.com
   Installing profile without version specified. Using version 1.7.5
   Installing profile with no configuration file

🌐 Installing Profile external-dns.community.tanzu.vmware.com
   Using profile version 0.10.0
   Using config external-values.yaml

🌐 Installing Profile app-toolkit.community.tanzu.vmware.com
   Installing profile without version specified. Using version 0.1.0
   Using config values.yaml
```

Bigger example installing app toolkit:
```
$ tanzu unmanaged-cluster create test --profile app-toolkit.projects.registry.vmware.com::values.yaml
...

🌐 Installing Profile app-toolkit.community.tanzu.vmware.com
   Installing profile without version specified. Using version 0.1.0
   Using config values.yaml

...

$ k get packageinstalls -A
NAMESPACE                   NAME                                     PACKAGE NAME                                          PACKAGE VERSION   DESCRIPTION           AGE
tanzu-package-repo-global   cartographer                             cartographer.community.tanzu.vmware.com               0.2.2             Reconcile succeeded   65s
tanzu-package-repo-global   cert-manager                             cert-manager.community.tanzu.vmware.com               1.6.1             Reconcile succeeded   99s
tanzu-package-repo-global   contour                                  contour.community.tanzu.vmware.com                    1.19.1            Reconcile succeeded   99s
tanzu-package-repo-global   fluxcd-source-controller                 fluxcd-source-controller.community.tanzu.vmware.com   0.21.2            Reconcile succeeded   99s
tanzu-package-repo-global   knative-serving                          knative-serving.community.tanzu.vmware.com            1.0.0             Reconcile succeeded   99s
tanzu-package-repo-global   kpack                                    kpack.community.tanzu.vmware.com                      0.5.1             Reconcile succeeded   99s
tkg-system                  app-toolkit.community.tanzu.vmware.com   app-toolkit.community.tanzu.vmware.com                0.1.0             Reconcile succeeded   109s
tkg-system                  cni                                      calico.community.tanzu.vmware.com                     3.22.1            Reconcile succeeded   109s


$ k get pods -A
NAMESPACE             NAME                                          READY   STATUS      RESTARTS   AGE
cartographer-system   cartographer-controller-86996bbb9-72gnz       1/1     Running     0          46s
cert-manager          cert-manager-56d644484c-8hz72                 1/1     Running     0          103s
cert-manager          cert-manager-cainjector-c77cb865-pwzm4        1/1     Running     0          103s
cert-manager          cert-manager-webhook-d7c4c8dc9-dsllw          1/1     Running     0          103s
flux-system           source-controller-5cfc8d5f94-jdn2g            1/1     Running     0          54s
knative-serving       activator-5d89dc486f-tmwgx                    1/1     Running     0          70s
knative-serving       autoscaler-7c7d7bb749-sz2k6                   1/1     Running     0          69s
knative-serving       controller-8c878f645-xd87v                    1/1     Running     0          70s
knative-serving       domain-mapping-544594f79b-5cs6v               1/1     Running     0          68s
knative-serving       domainmapping-webhook-5c56fcf66-5dk2s         1/1     Running     0          68s
knative-serving       net-certmanager-controller-8657b7f5dc-pxrn6   1/1     Running     0          70s
knative-serving       net-certmanager-webhook-765c755c57-l6lrx      1/1     Running     0          70s
knative-serving       net-contour-controller-59b89f9594-qhhzg       1/1     Running     0          70s
knative-serving       webhook-84679d6476-jxjst                      1/1     Running     0          68s
kpack                 kpack-controller-8b84485dc-pf7wh              1/1     Running     0          70s
kpack                 kpack-webhook-7674b5447-dttg5                 1/1     Running     0          70s
kube-system           calico-kube-controllers-6487584b54-tcsgt      1/1     Running     0          113s
kube-system           calico-node-tjvs2                             1/1     Running     0          113s
kube-system           coredns-78fcd69978-4x2qt                      1/1     Running     0          3m5s
kube-system           coredns-78fcd69978-qcljl                      1/1     Running     0          3m5s
kube-system           etcd-test-control-plane                       1/1     Running     0          3m21s
kube-system           kube-apiserver-test-control-plane             1/1     Running     0          3m18s
kube-system           kube-controller-manager-test-control-plane    1/1     Running     0          3m19s
kube-system           kube-proxy-jkkqx                              1/1     Running     0          3m5s
kube-system           kube-scheduler-test-control-plane             1/1     Running     0          3m18s
local-path-storage    local-path-provisioner-85494db59d-q6qdw       1/1     Running     0          3m5s
projectcontour        contour-6bc54547c4-5sclw                      1/1     Running     0          105s
projectcontour        contour-6bc54547c4-q7lv4                      1/1     Running     0          105s
projectcontour        contour-certgen-v1.19.1--1-gx458              0/1     Completed   0          105s
projectcontour        envoy-q9d97                                   2/2     Running     0          98s
tkg-system            kapp-controller-779d9777dc-qvnbk              1/1     Running     0          2m43s
```

Users can drop in the profile flag and mappings into the `config` command:
```
$ tanzu unmanaged-cluster config --profile fluent-bit:1.7.5:values.yaml test
Wrote configuration file to: test.yaml

$ cat test.yaml
ClusterName: test
KubeconfigPath: ""
ExistingClusterKubeconfig: ""
NodeImage: ""
Provider: kind
ProviderConfiguration: {}
Cni: calico
CniConfiguration: {}
PodCidr: 10.244.0.0/16
ServiceCidr: 10.96.0.0/16
TkrLocation: ""
AdditionalPackageRepos: []
PortsToForward: []
SkipPreflight: false
ControlPlaneNodeCount: "1"
WorkerNodeCount: "0"
Profiles:
  - name: fluent-bit
    config: values.yaml
    version: 1.7.5
```

jpmcb committed:
Implements unmanaged-cluster profiles

- Users can specify --profile via flag or config to automatically
  install a profile from an installed package repo.
- Users can set the --profile-version to install a specific version.
  This is optional and will default the the latest semver if not
  specified
- Users can set the --profile-config-file for a values yaml to be
  consumed by the installed profile

Signed-off-by: John McBride <jmcbride@vmware.com>

joshrosso reviewed (changes requested):
making this for my review, thanks!

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/semver/semver.go:
```diff
@@ -0,0 +1,317 @@
+//nolint
```
sweet :)

can we make a quick comment on why we copy (forked) this?

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/semver/semver.go:
```diff
@@ -0,0 +1,317 @@
+
+import "sort"
+
+// parsed returns the parsed form of a semantic version string.
```
sweet :)

can we make a quick comment on why we copy (forked) this?

(ie need to remove v requirement)

joshrosso replied:
That's good. I missed it!

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/packages/packages.go:
```diff
@@ -91,6 +91,10 @@ type PackageManager interface {
+	// GetRootServiceAccount returns the root service account
+	// given a namespace and the name of the service account
+	// Fails if service account doesn't exist
+	GetRootServiceAccount(ns, name string) (*v1.ServiceAccount, error)
```
nit: Probably best to return `nil` if a service account isn't found.

reasoning: if the service count isn't there it isn't an error of GetRootServiceAccount (like a failure to communicate with the cluster would be), it's an ok behavior and the caller should figure out if it wants to bubble up an error based on `nil` being returned.

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/cmd/create.go:
```diff
@@ -59,8 +62,9 @@ Exit codes are provided to enhance the automation of bootstrapping and are defin
-12 - Could not install CNI package.
-13 - Failed to merge kubeconfig and set context`
+12 - Could not install designated profile
+13 - Could not install CNI package.
```
For backwards compat, can we keep this one as `12` and add all the new ones above `12`?

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/cmd/create.go:
```diff
@@ -87,6 +91,9 @@ func init() {
 	CreateCmd.Flags().BoolVar(&co.skipPreflightChecks, "skip-preflight", false, "Skip the preflight checks; default is false")
 	CreateCmd.Flags().StringVar(&co.numContPlanes, "control-plane-node-count", "", "The number of control plane nodes to deploy; default is 1")
 	CreateCmd.Flags().StringVar(&co.numWorkers, "worker-node-count", "", "The number of worker nodes to deploy; default is 0")
+	CreateCmd.Flags().StringVar(&co.profile, "profile", "", "A profile to install. Should be the fully qualified package name found in the package repository")
```
At the beginning of this description, can we say `(experimental)`

joshrosso reviewed (commented):
@jpmcb as part of this PR, can you add to the reference documentation?

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/tanzu/tanzu.go:
```diff
@@ -252,23 +254,26 @@ func (t *UnmanagedCluster) Deploy(scConfig *config.UnmanagedClusterConfig) (int,
+	blockForRepoStatus(createdCoreRepo, pkgClient)
 
-	// Install the additional package repos
+	// Install and wait for the additional package repos
```
Let's only block for reconciliation when a profile is specified.

ie: Most bootstraps, which won't set a profile, shouldn't wait for additional package repos to reconcile.

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/tanzu/tanzu.go:
```diff
@@ -288,14 +293,41 @@ func (t *UnmanagedCluster) Deploy(scConfig *config.UnmanagedClusterConfig) (int,
+		}
+
+		if t.config.ProfileConfig == "" {
+			log.Style(outputIndent, color.FgYellow).Warnf("Installing profile with no configuration file\n")
```
Should this be a warning?

I wonder if it's expected and OK for some future metapackages to just install without config?

jpmcb replied:
Hmmmm good question.

My gut instinct is that _most_ packages will need at least some configuration and installing the default package without any values "works" ok but we should log a warning to the user that the package is using all default values and may not work 100% out of the box.

Even something [simple like `fluent-bit` needs](https://github.com/vmware-tanzu/community-edition/tree/main/addons/packages/fluent-bit/1.7.5) annotations configurations and stream sources to forward logs successfully.

---

Up to you, I don't know enough about the packaging expectations to make a call on this one

joshrosso replied:
Yeah, I hear you.

I'd hope most packages that could be used as profiles would have "sensible no-config defaults".

Either way we're guessing.

Let's keep the warning and see if it raising improper concern for users over time. :+1: 

_resolving_

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/tanzu/tanzu.go:
```diff
@@ -1025,10 +1057,83 @@ func mergeKubeconfigAndSetContext(mgr kubeconfig.Manager, kcPath string) error {
+// resolveProfilePkg picks the first package in the package repo
+// that matches the name and version of the provided profile
+// If the user did not specify a profile version, defaults to the first one found which should be the latest
+func resolveProfilePkg(mgr packages.PackageManager, profileName, profileVersion string) (*Package, error) {
```
Could we turn this into `resolvePkg` and re-use it for the resolveCNI logic too?

Seems like they are similar enough?

https://github.com/jpmcb/community-edition/blob/afe9592b4bbc4df82212f01432bb30ad1107d4d1/cli/cmd/plugin/unmanaged-cluster/tanzu/tanzu.go#L1136-L1160

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/cmd/create.go:
```diff
@@ -27,6 +27,9 @@ type createUnmanagedOpts struct {
 	portMapping               []string
 	numContPlanes             string
 	numWorkers                string
+	profile                   string
```
Perhaps as a stretch goal of this PR, I'd love to see profile accept multiple packages.

From a configuration standpoint, the API would likely be pretty simple, for example:

```yaml
profile:
- name: fluent-bit.community.tanzu.vmware.com
  config: values.yaml
  version: 1.7.5
```

But the complexity likely comes in from the command line flags. I wonder if a formatting convention would work?

```
--profile fluent-bit.community.tanzu.vmware.com:1.7.5:values.yaml,contour.community.tanzu.vmware.com:2.7.5:values2.yaml
```

In this model, `:` optionally separates package name from version and version from config. `,` separates multiple package entries.

The hope would be that most folks would be using the flag simply like:

```
--profile fluent-bit.community.tanzu.vmware.com
```

But we give the more complex flag option as well.

joshrosso reviewed (changes requested):
This is looking really good.

Made suggestions throughout.

jpmcb commented on cli/cmd/plugin/unmanaged-cluster/semver/semver.go:
```diff
@@ -0,0 +1,317 @@
+// license that can be found in the LICENSE file.
+//
+// Copyright 2022 VMware Tanzu Community Edition contributors. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
```
Question: do we also include our license header here? I put both, but i don't know the convention. Maybe @stmcginnis knows?

jpmcb replied:
Sounds good! That's what I've done and looks like the linters are happy!

jpmcb committed:
Implements multiple profiles

- Resets exit codes to be backwards compatible
- Updates comments / docs within code
- Makes resolving a package generic

Signed-off-by: John McBride <jmcbride@vmware.com>

jpmcb commented:
Pushed another commit that makes it possible to set multiple profiles - would be good to get a few eyes on this implementation!

This PR still needs a bit of work:
- [ ] Documentation updates
- [ ] Update the help messages in the CLI to reflect setting multiple profiles
- [ ] Write a few more tests for profile parsing
- [ ] Re-reviews of code implementations

jpmcb committed:
Documents unmanaged-cluster profiles

- Updates cobra usage function to wrap around terminals so long help
  decriptions don't break the displayed flags
- Adds additional tests to profile parsing

Signed-off-by: John McBride <jmcbride@vmware.com>

joshrosso commented:
> 



> @jpmcb @joshrosso I can't comment about code, but I have to say that I am not in favour of this implementation as it will lead to a lot of problems/issues coming from the community abusing the loose checks of this feature.
> 
> An example, from the testing John did.
> 
> ```
> tanzu unmanaged-cluster create test --profile external-dns.community.tanzu.vmware.com
> ```
> 
> This package will never be reconciled because it requires a value that is not provided https://github.com/vmware-tanzu/community-edition/tree/main/addons/packages/external-dns/0.10.0#external-dns-configuration

Heard and thanks for voicing this. We're still going to move ahead with this implementation and mark it as an experimental feature.

It will be on the user to know whether they need to include a values file.

In the future we'll aim to make this smarter, especially once our packages have YTT schemas where they declare their required values. This plumbing is still valid for that future state.

jpmcb commented:
> I have to say that I am not in favour of this implementation as it will lead to a lot of problems/issues coming from the community abusing the loose checks of this feature.

Totally heard. Unfortunately, today, we don't have a great way to "pre-check" that a values yaml file is needed or that a provided values yaml file matches the expected schema. 

This first pass will give users the ability to install and based on future user feedback, we can iterate and make this a better user experience.

joshrosso commented:
@jorgemoralespou could be good to get you sync'd up with @vkorbes and I if the primary contention is naming.

Let's get some time to talk it through, we have room to change the name prior to April 19th.

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/cmd/configure.go:
```diff
@@ -37,6 +37,9 @@ func init() {
 	ConfigureCmd.Flags().Bool("tty-disable", false, "Disable log stylization and emojis")
 	ConfigureCmd.Flags().StringSliceVar(&co.additionalRepo, "additional-repo", []string{}, "Addresses for additional package repositories to install")
+	ConfigureCmd.Flags().StringSliceVar(&co.profile, "profile", []string{}, "(experimental) A profile to install. May be specified multiple times. Should be the fully qualified package name or a prefix to a package name found in an installed package repository. Profile mappings supported - profile-name:profile-version:profile-config-file")
+	ConfigureCmd.Flags().StringSliceVar(&co.profileConfigPath, "profile-config-file", []string{}, "(experimental) Optional: path to a profile values yaml file. Uses default values (when available) if not provided. May be specified multiple times. Strings given via this flag are ordered in a queue and are enqueud in the order they are specified and dequeud when missing profile configs are encountered ...
```
typo in `enqueued` and `dequeued`.

@jpmcb is there an simpler way to say:

"Strings given via this flag are ordered in a queue and are enqueud in the order they are specified and dequeud when missing profile configs are encountered."

I feel like users might not have any clue what we're talking about.

jpmcb replied:
> I feel like users might not have any clue what we're talking about.

You're right, it is confusing. I couldn't really come up with a better way of saying "order won't matter when these flags are provided" and yet wanted to specify _how_ someone could order their flags.

This is partly a symptom of how cobra consumes these and actually is a long standing design problem in cobra around ordering flags.

A few thoughts:
- Just say "ordering doesn't work: use a config file"
- Recommend using config files or the full profile-mappings for multiple profiles
- Come up with a better way to describe this ordering.

jpmcb replied:
Planning to remove the references to queues: https://github.com/vmware-tanzu/community-edition/pull/3827#discussion_r840907023

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/cmd/configure.go:
```diff
@@ -37,6 +37,9 @@ func init() {
 	ConfigureCmd.Flags().StringSliceVar(&co.additionalRepo, "additional-repo", []string{}, "Addresses for additional package repositories to install")
+	ConfigureCmd.Flags().StringSliceVar(&co.profile, "profile", []string{}, "(experimental) A profile to install. May be specified multiple times. Should be the fully qualified package name or a prefix to a package name found in an installed package repository. Profile mappings supported - profile-name:profile-version:profile-config-file")
+	ConfigureCmd.Flags().StringSliceVar(&co.profileConfigPath, "profile-config-file", []string{}, "(experimental) Optional: path to a profile values yaml file. Uses default values (when available) if not provided. May be specified multiple times. Strings given via this flag are ordered in a queue and are enqueud in the order they are specified and dequeud when missing profile configs are encountered ...
+	ConfigureCmd.Flags().StringSliceVar(&co.profileVersion, "profile-version", []string{}, "(experimental) Optional: the version of a profile to install. Uses the latest version if not provided. May be specified multiple times. Installs latest if not provided. May be specified multiple times. Values specified via this flag are ordered in a queue and are enqueud in the order they are specified and de ...
```
typo in `enqueued` and `dequeued`.

@jpmcb is there an simpler way to say:

"Strings given via this flag are ordered in a queue and are enqueud in the order they are specified and dequeud when missing profile configs are encountered."

I feel like users might not have any clue what we're talking about.

jpmcb replied:
Planning to remove the references to queues: https://github.com/vmware-tanzu/community-edition/pull/3827#discussion_r840907023

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/tanzu/tanzu.go:
```diff
@@ -1025,31 +1060,84 @@ func mergeKubeconfigAndSetContext(mgr kubeconfig.Manager, kcPath string) error {
-		return nil, err
+		log.Errorf("failed to get root service account: %s\n", err.Error())
+		return err
 	}
```
Below here, we need to check if `rootSvcAcct` is nil, if it is, return an error.

Otherwise, should that condition ever happen, we'll panic below when you call `rootSvcAcct.Name`.

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/cmd/configure.go:
```diff
@@ -37,6 +37,9 @@ func init() {
 	ConfigureCmd.Flags().Bool("tty-disable", false, "Disable log stylization and emojis")
 	ConfigureCmd.Flags().StringSliceVar(&co.additionalRepo, "additional-repo", []string{}, "Addresses for additional package repositories to install")
+	ConfigureCmd.Flags().StringSliceVar(&co.profile, "profile", []string{}, "(experimental) A profile to install. May be specified multiple times. Should be the fully qualified package name or a prefix to a package name found in an installed package repository. Profile mappings supported - profile-name:profile-version:profile-config-file")
+	ConfigureCmd.Flags().StringSliceVar(&co.profileConfigPath, "profile-config-file", []string{}, "(experimental) Optional: path to a profile values yaml file. Uses default values (when available) if not provided. May be specified multiple times. Strings given via this flag are ordered in a queue and are enqueud in the order they are specified and dequeud when missing profile configs are encountered ...
```
@jpmcb As I read this, your implementation supports modes of specifying a profile.

```
--profile fluent-bit:1.7.5:path-to-my-config.yaml
```

and

```
--profile fluent-bit --profile-version 1.7.5 --profile-config-file path-to-my-config.yaml
```

If i'm reading the parsing code correctly, these 2 create equivalent outcomes.

I wonder if we're better offering the `--profile` option exclusively.

My initial reasoning is:

1. Unless it's an extremely high-value UX benefit, it might not make sense to have 2 ways to do the same thing.
2. Code complexity is increased a non-trivial amount by satisfying these 2 models.

What are your thoughts? And/or am I overlooking anything?

jpmcb replied:
Hmmm good question. Yes, essentially, this UX supports both profile mapping and just the flag. What's unique about this is it also supports mixing those.

So a user could do something like:
```
--profile fluent-bit:1.7.5
--profile-config-file my-config.yaml
```
I think this is a powerful UX that would allow people to drop in and replace single fields in profile mappings like in the above example: the config file is explicitly declared instead of the profile mapping

But you're right: there is increased code complexity and possibly increased effect of user error/unintended behavior. I'm a bit torn because I really like the flexability of having both and being able to mix and match.

jpmcb replied:
Chatted, and we definitely want to go with the simpler UX of a single `--profile` flag with profile mappings:

```
--profile fluent-bit:1.7.5:path-to-values.yaml
```

If users don't want to include a version but they do want to include a values yaml file, they can do either:

```
--profile fluent-bit::path-to-values.yaml
```

or we reserve the `latest` keyword to resolve to the latest package found (the default behavior if no version specified)

```
--profile fluent-bit:latest:path-to-values.yaml
```

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/tanzu/tanzu.go:
```diff
@@ -1025,31 +1060,84 @@ func mergeKubeconfigAndSetContext(mgr kubeconfig.Manager, kcPath string) error {
 	}
+
+	if profilePkg.fqPkgName == "" {
+		return nil, fmt.Errorf("no package was resolved for named %s with version %s", pkgName, pkgVersion)
```
typo: `name`

joshrosso commented on docs/site/content/docs/ref-unmanaged-cluster.md:
````diff
@@ -48,6 +48,104 @@ using the default `kind` provider
+run the following
+
+```sh
+tanzu unmanaged-cluster create my-cluster --profile fluent-bit
````
We need to use the package's FQN, right? If so, might be good to update these examples with that.

jpmcb replied:
Not necessarily: users can provide the prefix to a package FQN (like we do when installing the CNI).

So running the above works and selects the `fluent-bit.community.tanzu.vmware.com` package

joshrosso commented:
@jpmcb question (i'm not sure)

should we fully fail when a profile doesn't install?

```
$ tanzu unmanaged-cluster create my-cluster3 --tkr=projects.registry.vmware.com/tce/tkr:v0.17.0 --profile fluent-bit.community.tanzu.vmware.com

📁 Created cluster directory

🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Compatibility file exists at /home/josh/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v3
   Custom TKr projects.registry.vmware.com/tce/tkr:v0.17.0 NOT found in compatibility file. Proceed with caution, the provided TKr may not work with this version of unmanaged-cluster

🔧 Resolving TKr
   projects.registry.vmware.com/tce/tkr:v0.17.0
   TKr exists at /home/josh/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v0.17.0
   Rendered Config: /home/josh/.config/tanzu/tkg/unmanaged/my-cluster3/config.yaml
   Bootstrap Logs: /home/josh/.config/tanzu/tkg/unmanaged/my-cluster3/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind:v1.22.4

📦 Selected core package repository
   projects.registry.vmware.com/tce/repo-10:0.10.0

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tce/kapp-controller-multi-pkg:v0.30.1

🚀 Creating cluster my-cluster3
   Cluster creation using kind!
   ❤️  Checkout this awesome project at https://kind.sigs.k8s.io
   Base image downloaded
   Cluster created
   To troubleshoot, use:

   kubectl ${COMMAND} --kubeconfig /home/josh/.config/tanzu/tkg/unmanaged/my-cluster3/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   tkg-core-repository package repo status: Reconcile succeeded

🌐 Installing CNI
   calico.community.tanzu.vmware.com:3.22.1

🌐 Installing Profile fluent-bit.community.tanzu.vmware.com
could not resolve profile. Error: no package was resolved for named fluent-bit.community.tanzu.vmware.com with version
Error: exit status 14
```

I wonder if, similar to CNI, we'd be better off printing a warning and still letting the cluster go through with the rest of its bootstrap.

joshrosso reviewed (changes requested):
@jpmcb a few more things to address, but overall this thing is look excellent in my local testing.

jpmcb commented:
> should we fully fail when a profile doesn't install? ... I wonder if, similar to CNI, we'd be better off printing a warning and still letting the cluster go through with the rest of its bootstrap

@joshrosso - I agree, I think we should probably hand the cluster back over to the user (like we do with CNI) and not fail fully. Good catch.

jorgemoralespou commented:
How will the user, if he gota cluster back that some packages were not installed successfully? Will that be part of the status or output?
Is this the behavior you would expect in the tools you use?

joshrosso commented:
> How will the user, if he gota cluster back that some packages were not installed successfully? Will that be part of the status or output?

It will not. Similar to CNI, we will **not** block the install on the reconciliation of `PackageInstall`->`App`.

We are open to considering this out-of-band from this work in a separate issue.

>  Is this the behavior you would expect in the tools you use?

Depends on the tool. Every time I do a `kubectl apply` I expect this exact behavior

jpmcb committed:
Simplifies installing profiles

- Users can install profiles via the `--profile` flag during
  unmanaged-cluster creation
- Removes `--profile-version` and `--profile-config-file` flags

Signed-off-by: John McBride <jmcbride@vmware.com>

joshrosso reviewed (dismissed):
thanks for addressing all the suggestions.

also works great for me.

stmcginnis commented:
Looks like we may have a regression in cluster deletion. If the cluster fails to be created, that normally creates the cluster metadata folder under `~/.config/tanzu/tkg/unmanaged/[cluster-name]`. That does still happen.

The difference is normally we can then call `tanzu uc rm` and it will delete the metadata directory. When doing that after failing to create a cluster with this change, I now get:

```txt
$ ./artifacts/linux/amd64/cli/unmanaged-cluster/v0.12.0-dev.1/tanzu-unmanaged-cluster-linux_amd64 create foo --profile=kpack.community.tanzu.vmware.com:0.5.1
directory /home/smcginnis/.config/tanzu/tkg/unmanaged/foo already exists, this cluster must be deleted before proceeding
~/src/vmware-tanzu/community-edition (unmanaged/profiles)
$ ./artifacts/linux/amd64/cli/unmanaged-cluster/v0.12.0-dev.1/tanzu-unmanaged-cluster-linux_amd64 rm foo

🧪 Deleting cluster: foo
Failed to delete cluster. Error: failed to locate a config file at /home/smcginnis/.config/tanzu/tkg/unmanaged/foo/config.yaml
```

It _should_ try to get information, if it can't get enough to delete the cluster assume it is gone, then remove the metadata folder.

Not sure if that is directly caused by any of these changes, or if the failure condition this is hitting is different than most I've encountered before.

stmcginnis commented:
> should we fully fail when a profile doesn't install?

I think we actually should. This functionality makes it less likely that a user is interactively running this and helps enable the use case of scripting the deployment of a tanzu cluster in a pipeline or something like that.

In this case, the only way they have to validate that the deployment was successful is to inspect the return code from calling `create`. If they are saying they want a cluster created with a certain set of packages, and we return 0, they continue on thinking everything worked and they have the environment they expect to have for their given scenario. Then if they try to use this cluster to deploy a workload, it won't be immediately clear why it is failing if the cluster environment they are deploying isn't actually configured the way they expect it to be.

stmcginnis commented:
Trying to test locally and I can't get past this error:

```txt
🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Downloaded to: /home/smcginnis/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v3             
failed parsing the TKr compatibility file. Error: could not find compatible CLI version in compatibility file
```

joshrosso commented:
> Looks like we may have a regression in cluster deletion. If the cluster fails to be created, that normally creates the cluster metadata folder under `~/.config/tanzu/tkg/unmanaged/[cluster-name]`. That does still happen.
> 
> The difference is normally we can then call `tanzu uc rm` and it will delete the metadata directory. When doing that after failing to create a cluster with this change, I now get:
> 
> ```
> $ ./artifacts/linux/amd64/cli/unmanaged-cluster/v0.12.0-dev.1/tanzu-unmanaged-cluster-linux_amd64 create foo --profile=kpack.community.tanzu.vmware.com:0.5.1
> directory /home/smcginnis/.config/tanzu/tkg/unmanaged/foo already exists, this cluster must be deleted before proceeding
> ~/src/vmware-tanzu/community-edition (unmanaged/profiles)
> $ ./artifacts/linux/amd64/cli/unmanaged-cluster/v0.12.0-dev.1/tanzu-unmanaged-cluster-linux_amd64 rm foo
> 
> 🧪 Deleting cluster: foo
> Failed to delete cluster. Error: failed to locate a config file at /home/smcginnis/.config/tanzu/tkg/unmanaged/foo/config.yaml
> ```
> 
> It _should_ try to get information, if it can't get enough to delete the cluster assume it is gone, then remove the metadata folder.
> 
> Not sure if that is directly caused by any of these changes, or if the failure condition this is hitting is different than most I've encountered before.

I don't believe this is a regression produced by this PR. I have requested it's fixed as part of the `minikube` provider work.

https://github.com/vmware-tanzu/community-edition/pull/3674#issuecomment-1086229724

joshrosso commented:
> > should we fully fail when a profile doesn't install?
> 
> I think we actually should. This functionality makes it less likely that a user is interactively running this and helps enable the use case of scripting the deployment of a tanzu cluster in a pipeline or something like that.
> 
> In this case, the only way they have to validate that the deployment was successful is to inspect the return code from calling `create`. If they are saying they want a cluster created with a certain set of packages, and we return 0, they continue on thinking everything worked and they have the environment they expect to have for their given scenario. Then if they try to use this cluster to deploy a workload, it won't be immediately clear why it is failing if the cluster environment they are deploying isn't actually configured the way they expect it to be.

Thanks for the input. At this point, I'm not aligned, because I feel as though we're taking a UX hit for the sake of having exhaustive error codes. Open to reconsidering blocking and validating package installs in the context of a separate issue. If we do that, we need to start doing the same for CNI, which we don't do today.

stmcginnis commented:
OK, things are looking good now:

```
$ tanzu uc create test --tkr projects.registry.vmware.com/tce/tkr:v0.17.0-dev-1 --profile fluent-bit.community.tanzu.vmware.com

📁 Created cluster directory

🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Compatibility file exists at /home/smcginnis/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v3
   Custom TKr projects.registry.vmware.com/tce/tkr:v0.17.0-dev-1 NOT found in compatibility file. Proceed with caution, the provided TKr may not work with this version of unmanaged-cluster

🔧 Resolving TKr
   projects.registry.vmware.com/tce/tkr:v0.17.0-dev-1
   TKr exists at /home/smcginnis/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v0.17.0-dev-1
   Rendered Config: /home/smcginnis/.config/tanzu/tkg/unmanaged/test/config.yaml
   Bootstrap Logs: /home/smcginnis/.config/tanzu/tkg/unmanaged/test/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind:v1.22.4

📦 Selected core package repository
   projects.registry.vmware.com/tce/repo-10:0.10.0

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.11.0

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tce/kapp-controller-multi-pkg:v0.30.1

🚀 Creating cluster test
   Cluster creation using kind!
   ❤️  Checkout this awesome project at https://kind.sigs.k8s.io
   Base image downloaded             
   Cluster created                 
   To troubleshoot, use:

   kubectl ${COMMAND} --kubeconfig /home/smcginnis/.config/tanzu/tkg/unmanaged/test/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running                    

📧 Installing package repositories
   tkg-core-repository package repo status: Reconcile succeeded             
   projects.registry.vmware.com-tce-main-0.11.0 package repo status: Reconcile succeeded             

🌐 Installing CNI
   calico.community.tanzu.vmware.com:3.22.1

🌐 Installing Profile fluent-bit.community.tanzu.vmware.com
   Selected package fluent-bit.community.tanzu.vmware.com
   Installing profile without version specified. Using version 1.7.5
   Installing profile with no configuration file

✅ Cluster created

🎮 kubectl context set to test

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete test
```

```
$ k get po -A
NAMESPACE            NAME                                         READY   STATUS    RESTARTS   AGE
fluent-bit           fluent-bit-g5bvb                             1/1     Running   0          58s
kube-system          calico-kube-controllers-5757f95646-hgs4t     1/1     Running   0          57s
kube-system          calico-node-5ch9g                            1/1     Running   0          57s
kube-system          coredns-78fcd69978-klx89                     1/1     Running   0          102s
kube-system          coredns-78fcd69978-xhcpn                     1/1     Running   0          102s
kube-system          etcd-test-control-plane                      1/1     Running   0          116s
kube-system          kube-apiserver-test-control-plane            1/1     Running   0          115s
kube-system          kube-controller-manager-test-control-plane   1/1     Running   0          116s
kube-system          kube-proxy-frkbj                             1/1     Running   0          102s
kube-system          kube-scheduler-test-control-plane            1/1     Running   0          116s
local-path-storage   local-path-provisioner-85494db59d-4fjnn      1/1     Running   0          102s
tkg-system           kapp-controller-779d9777dc-vfnvd             1/1     Running   0          102s
```

```
$ k get app -A
NAMESPACE    NAME                                    DESCRIPTION           SINCE-DEPLOY   AGE
tkg-system   cni                                     Reconcile succeeded   70s            78s
tkg-system   fluent-bit.community.tanzu.vmware.com   Reconcile succeeded   68s            77s
```

stmcginnis commented on cli/cmd/plugin/unmanaged-cluster/cmd/utils.go:
```diff
@@ -8,7 +8,10 @@ import (
 
+	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
+
```
nit: Not sure why the lint check didn't complain, but there shouldn't be a separation here. `github.com/spf` and `golang.org/x` are both third party imports, so should be part of the same import grouping.

stmcginnis reviewed (dismissed):
Looks good, and locally testing appears to be working fine. Really like the improvided cobra help output.

jpmcb commented:
> 🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Downloaded to: /home/smcginnis/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v3             
failed parsing the TKr compatibility file. Error: could not find compatible CLI version in compatibility file

Hmmm what version do you get when you run `tanzu um version`? This is the top of the v3 compatibility file:

```
version: v1
unmanagedClusterPluginVersions:
- version: dev
  supportedTkrVersions:
  - image: projects.registry.vmware.com/tce/tkr:v0.17.0-dev-1
```

Might be because I've been running the binary directly which gets `dev` as the version:

```
$ ./unmanaged-cluster version
dev
```

but yes, 💯 to what josh said on specifying the `--tkr` flag!

jpmcb commented:
@joshrosso @stmcginnis - addressed a nit and ready for another review/merge 👀

stmcginnis commented:
> Hmmm what version do you get when you run tanzu um version?

I initially was running that too:

```
$ ./artifacts/linux/amd64/cli/unmanaged-cluster/v0.12.0-dev.1/tanzu-unmanaged-cluster-linux_amd64 version
v0.12.0-dev.1
```

The comments on #3895 sound good to me.

stmcginnis reviewed (approved):
Thanks!

vkorbes commented:
Necromancing this to pick up the naming discussion.

We want something that means "packages that get installed during cluster setup." On a scale from descriptive to facetious, suggestions:

- `--install-packages`
- `--setup-packages`
- `--pre-install`
- `--setup`
- `--preload`
- `--bootstrap`
- `--kickoff`

We're picking one by EOD tomorrow (Apr 14), please do chime in with your preferences or other suggestions. Thanks!

stmcginnis commented:
My vote would be `--preload` or `--preinstall`, in that order.

vrabbi commented:
My vote would be `--install-packages` or `--pre-install` in that order

jpmcb commented:
I don't mind for any of these 3: `--preload`, `--pre-install`, `--install-packages`

swalner-vmware commented:
I would go for `--pre-install-packages`

joshrosso commented:
For a potential interim solution, my preference is `--install-packages`.

My primary tiny criticism for options like `--preload` are that they feel just as ambiguous as `--profile`.

jorgemoralespou commented:
From those in the list, `--install-packages` is my vote

vrabbi commented:
I think --preload is not clear in what it means and doesnt indicate something would be installed rather it indicates something would be imported into a cluster like kind has the idea of preloaded images but they are just added to the cri level and not run in the cluster

garrying commented:
+1 to `--install-packages`

kartiklunkad26 commented:
+1 for `--install-packages`

thesteve0 commented:
+1 `--install-packages` with the hopes that also includes meta-packages

jorgemoralespou commented:
@thesteve0 isn't a meta-package just another package?

vrabbi commented:
A meta package is simply a package that installs other packaves and in yesterdays demo app toolkit was installed this way so shouldnt be an issue at all

vkorbes commented:
As per the majority of votes, we're going with `--install-packages`. Thanks so much everyone who chimed in!
